File: week4/2sum.rb
class SearchIndex

  def initialize data
    @index ={}
    @data = data
    for number in @data
       @index[number] = true
    end
  end

  def look_up t  
    for x in @data do
      y = t - x
      if (@index[y] && x != y) 
        return true
      end
    end  
    return false
  end
end

class BruteForce
  def look_up t  
    for x in @data do
      for y in @data do      
        if ( x + y == t && x != y) 
          return true
        end
      end
    end  
    return false
  end
end

def load_file file_path
  input = []
  File.open(file_path, "r") do |f|
    f.each_line do |line|
      input.push(line.to_i) 
    end
  end
  return input
end

def main
  start = Time.now
  puts "Starting algorithm"
 
  input = load_file("simple-data1.txt")
  puts "File read " +  (Time.now - start).to_s
 
  index = SearchIndex.new(input)
  puts "Index created " +  (Time.now - start).to_s

  counter = 0
  (-10000..10000).each do |t|
    counter += 1 if index.look_up(t)
  end
  puts "Numbers in [-10000, 10000] calculated in " +  (Time.now - start).to_s
  puts counter
end

main

File: week4/cpp/2sum.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <numeric>
#include <sstream>
#include <string>
#include <tuple>
#include <unordered_map>
#include <vector>

class SearchIndex
{
public:
    SearchIndex( std::vector< int64_t > data )
    {
        m_data = std::move( data );
		std::sort(m_data.begin(), m_data.end());

        for ( int64_t x : m_data )
        {
            m_index[ x ] = true;
        }
    }

    bool
    look_up( int64_t t )
    {
        for ( int64_t x : m_data )
        {
            int64_t y = t - x;
            auto it = m_index.find( y );
            if ( it != m_index.end( ) && x != y )
            {
                return true;
            }
        }
        return false;
    }

private:
    std::unordered_map< int64_t, bool > m_index;
    std::vector< int64_t > m_data;
};

std::vector< int64_t >
read_file( const std::string& file_path )
{
    std::ifstream infile( file_path );
    std::string line;
    std::vector< int64_t > data;
    data.reserve( 1000000 );
    while ( std::getline( infile, line ) )
    {
        std::istringstream iss( line );
        int64_t n;
        if ( !( iss >> n ) )
        {
            break;
        }  // error
        data.push_back( n );
    }
    return data;
}

int
main( )
{
    SearchIndex index( read_file( "hw.txt" ) );

    int32_t counter = 0;
    for ( int64_t t = -10000; t <= 10000; t++ )
    {
        if ( index.look_up( t ) )
        {
            counter++;
        }
    }

    std::cout << counter;
    std::string s;
    std::cin >> s;
    return 0;
}

